cpp/src/arrow/CMakeLists.txt

```diff
@@ -26,6 +26,200 @@ add_dependencies(arrow-all
+  endif()
+  list(APPEND ARROW_STATIC_INSTALL_INTERFACE_LIBS CURL::libcurl)
+endif()
+
+if(ARROW_WITH_UTF8PROC)
+  if(utf8proc_SOURCE STREQUAL "SYSTEM")
+    list(APPEND ARROW_STATIC_INSTALL_INTERFACE_LIBS utf8proc::utf8proc)
+  endif()
+endif()
+
+if(ARROW_WITH_RE2)
+  if(re2_SOURCE STREQUAL "SYSTEM")
+    list(APPEND ARROW_STATIC_INSTALL_INTERFACE_LIBS re2::re2)
+  endif()
+endif()
+
+# This should be done after if(ARROW_ORC) and if(ARROW_WITH_OPENTELEMETRY)
+# because they depend on Protobuf.
+if(ARROW_WITH_PROTOBUF)
+  if(Protobuf_SOURCE STREQUAL "SYSTEM")
+    list(APPEND ARROW_STATIC_INSTALL_INTERFACE_LIBS ${ARROW_PROTOBUF_LIBPROTOBUF})
+  endif()
+endif()
+
+if(ARROW_ENABLE_THREADING)
+  list(APPEND ARROW_SHARED_PRIVATE_LINK_LIBS Threads::Threads)
+  list(APPEND ARROW_STATIC_LINK_LIBS Threads::Threads)
+endif()
+
+if(NOT MSVC_TOOLCHAIN)
+  list(APPEND ARROW_SHARED_INSTALL_INTERFACE_LIBS ${CMAKE_DL_LIBS})
+  list(APPEND ARROW_STATIC_INSTALL_INTERFACE_LIBS ${CMAKE_DL_LIBS})
+endif()
+
+set(ARROW_TEST_LINK_TOOLCHAIN arrow::flatbuffers ${ARROW_GTEST_GMOCK}
+                              ${ARROW_GTEST_GTEST_MAIN})
+set(ARROW_TEST_STATIC_LINK_LIBS arrow_testing_static arrow_static
+                                ${ARROW_TEST_LINK_TOOLCHAIN})
+set(ARROW_TEST_SHARED_LINK_LIBS arrow_testing_shared arrow_shared
+                                ${ARROW_TEST_LINK_TOOLCHAIN})
+if(NOT MSVC)
+  list(APPEND ARROW_TEST_SHARED_LINK_LIBS ${CMAKE_DL_LIBS})
+endif()
+
+# ----------------------------------------------------------------------
+# Handle platform-related libraries like -pthread
+
+set(ARROW_SYSTEM_LINK_LIBS)
+
+if(CMAKE_THREAD_LIBS_INIT)
+  string(APPEND ARROW_PC_LIBS_PRIVATE " ${CMAKE_THREAD_LIBS_INIT}")
+endif()
+
+if(WIN32)
+  list(APPEND ARROW_SYSTEM_LINK_LIBS "ws2_32.dll")
+endif()
+
+if(NOT WIN32 AND NOT APPLE)
+  # Pass -lrt on Linux only
+  list(APPEND ARROW_SYSTEM_LINK_LIBS rt)
+endif()
+
+list(APPEND ARROW_SHARED_INSTALL_INTERFACE_LIBS ${ARROW_SYSTEM_LINK_LIBS})
+list(APPEND ARROW_STATIC_INSTALL_INTERFACE_LIBS ${ARROW_SYSTEM_LINK_LIBS})
+
+# Need -latomic on Raspbian.
+# See also: https://issues.apache.org/jira/browse/ARROW-12860
+if(${CMAKE_SYSTEM_NAME} STREQUAL "Linux" AND ${CMAKE_SYSTEM_PROCESSOR} MATCHES "armv7")
+  string(APPEND ARROW_PC_LIBS_PRIVATE " -latomic")
+  list(APPEND ARROW_SHARED_INSTALL_INTERFACE_LIBS "atomic")
+  list(APPEND ARROW_STATIC_INSTALL_INTERFACE_LIBS "atomic")
+endif()
+
+function(arrow_add_object_library PREFIX)
+  set(SOURCES ${ARGN})
+  string(TOLOWER "${PREFIX}" prefix)
+  if(WIN32)
+    add_library(${prefix}_shared OBJECT ${SOURCES})
+    add_library(${prefix}_static OBJECT ${SOURCES})
+    set_target_properties(${prefix}_shared PROPERTIES POSITION_INDEPENDENT_CODE ON)
+    set_target_properties(${prefix}_static PROPERTIES POSITION_INDEPENDENT_CODE ON)
+    target_compile_definitions(${prefix}_shared PRIVATE ARROW_EXPORTING)
+    target_compile_definitions(${prefix}_static PRIVATE ARROW_STATIC)
```

Comment:
Hmm, how does this work if we're compiling e.g. libarrow_acero.dll? Am I missing something?

Reply:
`arrow_add_object_library()` is only used for `libarrow.dll` related source files. It's not used for `libarrow_acero.dll` and others. So we can use `ARROW_EXPORTING`/`ARROW_STATIC` here. (We don't use `ARROW_ACERO_EXPORTING`/`ARROW_ACERO_STATIC` and so on here.)